src/magic_folder/scanner.py

```diff
@@ -0,0 +1,147 @@
+    implementer,
+)
+from twisted.python.filepath import (
+    FilePath,
+)
+from twisted.internet.defer import (
+    inlineCallbacks,
+    returnValue,
+)
+from twisted.internet.task import (
+    deferLater,
+    react,
+)
+from twisted.application.internet import (
+    TimerService,
+)
+from twisted.application.service import (
+    IService,
+)
+from eliot import (
+    start_action,
+)
+
+from .magicpath import (
+    path2magic,
+)
+
+
+@attr.s
+@implementer(IService)
+class ScannerService(TimerService):
+    """
+    Periodically scan a local Magic Folder for new or updated files
+    """
+    _reactor = attr.ib()
+    _config = attr.ib()
+    _local_snapshot_service = attr.ib()
+    _status = attr.ib()
+
+    def __attrs_post_init__(self):
+        # XXX do we want a "scan interval" too?
+        TimerService.__init__(
+            self,
+            self._config.poll_interval,
+            self._scan,
+        )
+
+    @inlineCallbacks
+    def _scan(self):
+        """
+        Perform a scan for new files.
+        """
+        # XXX probably want a lock ("or something") so we don't do
+        # overlapping scans (i.e. if a scan takes longer than the
+        # poll_interval we should not start a second one)
+        with start_action(action_type="scanner:find-updates") as action:
+            duration = yield find_updated_files(self._reactor, self._config, self._modified_file)
+            action.add_success_fields(scan_duration=duration)
+        # XXX update/use IStatus to report scan start/end
+
+    def _modified_file(self, path):
+        """
+        Internal helper.
+        Called when we find a new or modified file.
+        """
+        d = self._local_snapshot_service.add_file(path)
+
+        def bad(f):
+            # might want to expose some errors to users / status
+            print(f)
+        d.addErrback(bad)
+
+
+def _is_newer_than_current(folder_config, name, local_mtime):
+    """
+    Determine if `local_mtime` is newer than the existing Snapshot for
+    `name`. If there is no existing Snapshot for the name then True is
+    returned.
+
+    :param folder_config: our configuration
+    :param unicode name: the mangled name of the Snapshot
+    :param int local_mtime: timestamp of the current local file, in seconds.
+    """
+    # if we have a LocalSnapshot we've already queued up some changes
+    try:
+        localsnap = folder_config.get_local_snapshot(name)
+        existing_mtime = localsnap.metadata["mtime"]
+    except KeyError:
+        existing_mtime = None
+
+    # if we have no LocalSnapshot(s) proceed to see if we have a
+    # remotesnapshot recorded
+    if existing_mtime is None:
+        try:
+            existing_mtime = folder_config.get_remotesnapshot_mtime(name)
+        except KeyError:
+            existing_mtime = None
+
+    if existing_mtime is None:
+        # we have no record of this file; it must be new.
+        return True
+    return local_mtime > existing_mtime
```

Comment:
We want to compare equality here, not if the local time is newer. If they are not equal, then we should upload, even if the on-disk is older than we recorded.

Reply:
it's not clear to me which one is right, so okay .. changed